Data/IDProvider.cs

```diff
@@ -0,0 +1,84 @@
+﻿using System;
+using System.Diagnostics;
+using Anvil.CSharp.Logging;
+
+namespace Anvil.CSharp.Data
+{
+    /// <summary>
+    /// An instance that provides a unique IDs each time requested.
+    /// Includes a mechanism to detect when ID supply is near exhaustion.
+    /// </summary>
+    public class IDProvider
+    {
+        /// <summary>
+        /// Triggers the first time that any instance of <see cref="IDProvider"/> passes its
+        /// <see cref="SupplyWarningThreshold"/> for the first time.
+        /// This gives the application the opportunity to react before IDs are exhausted.
+        /// </summary>
+        public static event Action<IDProvider> OnIDLimitGlobalWarning;
+
+        /// <summary>
+        /// A reserved ID to represent "No ID".
+        /// <see cref="GetNextID"/> will never return this value even when IDs roll over and are no longer unique.
+        /// </summary>
+        public const int UNSET_ID = 0;
+
+        /// <summary>
+        /// Triggers the first time that <see cref="SupplyWarningThreshold"/> is passed.
+        /// This gives the consumer the opportunity to react before IDs are exhausted.
+        /// </summary>
+        public event Action OnIDLimitWarning;
+
+        public readonly uint SupplyWarningThreshold;
+
+        private bool m_HasIDWarningTriggered = false;
+        private uint m_LastNewID = 0;
+
+
+        /// <summary>
+        /// Creates a new ID provider optionally allowing a supply warning threshold to be set.
+        /// </summary>
+        /// <param name="supplyWarningThreshold">
+        /// (default: <see cref="uint.MaxValue"/>) The threshold which, when passed, triggers
+        /// <see cref="OnIDLimitImminent"/>.
+        /// </param>
+        public IDProvider(uint supplyWarningThreshold = uint.MaxValue-1)
+        {
+            SupplyWarningThreshold = supplyWarningThreshold;
+        }
+
+        /// <summary>
+        /// Provide an unused ID.
+        /// </summary>
+        /// <returns>A unique ID.</returns>
+        /// <remarks>
+        /// If <see cref="uint"/> has rolled over an error will be logged and IDs are no longer guaranteed to be unique.
+        /// </remarks>
+        public uint GetNextID()
+        {
+            uint id = ++m_LastNewID;
+            CheckIfIDThresholdCrossed();
+
+            if (m_LastNewID == 0)
+            {
+                Log.GetLogger(this).Error($"{nameof(IDProvider)} has passed its maximum ID value. IDs provided are no longer guaranteed to be unique.");
+                // Push the ID past the UNSET_ID value so at least THAT remains unique.
+                id = ++m_LastNewID;
+            }
```

Comment:
Does this work? I thought you had to use the `checked` or `unchecked` when dealing with overflow?

Reply:
Using those keywords overrides the compiler setting. By default Unity doesn't have arithmetic overflow checking enabled. Even with Editor Debug mode enabled. I'm not sure how you turn on checking application wide in Unity, it is a compiler argument that you'd normally enable through project preferences in the IDE.

(Isn't enabled in .Net Fiddle either https://dotnetfiddle.net/58H0eP)

I'm thinking we leave it as is. If somebody turns on arithmetic checking then they'll get an exception. Since the overflow of the ID is an error case, that we've just provided graceful handling for, it makes sense to let it throw an exception if the developer wanted the safety of overflow checking.

Reply:
Yep, ok this makes sense.